filters/log/log.go

```diff
@@ -0,0 +1,145 @@
+package log
+
+import (
+	"bytes"
+	"encoding/json"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"os"
+
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	AuditLogName        = "auditLog"
+	AuthUserKey         = "auth-user"
+	AuthRejectReasonKey = "auth-reject-reason"
+)
+
+type auditLog struct {
+	writer     io.Writer
+	maxBodyLog int
+}
+
+type teeBody struct {
+	body      io.ReadCloser
+	buffer    *bytes.Buffer
+	teeReader io.Reader
+	maxTee    int
+}
+
+type auditDoc struct {
+	Method      string         `json:"method"`
+	Path        string         `json:"path"`
+	Status      int            `json:"status"`
+	AuthStatus  *authStatusDoc `json:"authStatus,omitempty"`
+	RequestBody string         `json:"requestBody,omitempty"`
+}
+
+type authStatusDoc struct {
+	User     string `json:"user,omitempty"`
+	Rejected bool   `json:"rejected"`
+	Reason   string `json:"reason,omitempty"`
+}
+
+func newTeeBody(rc io.ReadCloser, maxTee int) io.ReadCloser {
+	b := bytes.NewBuffer(nil)
+	tb := &teeBody{
+		body:   rc,
+		buffer: b,
+		maxTee: maxTee}
+	tb.teeReader = io.TeeReader(rc, tb)
+	return tb
+}
+
+func (tb *teeBody) Read(b []byte) (int, error) { return tb.teeReader.Read(b) }
+func (tb *teeBody) Close() error               { return tb.body.Close() }
+
+func (tb *teeBody) Write(b []byte) (int, error) {
+	if tb.maxTee < 0 {
+		return tb.buffer.Write(b)
+	}
+
+	wl := len(b)
+	if wl >= tb.maxTee {
+		wl = tb.maxTee
+	}
+
+	n, err := tb.buffer.Write(b[:wl])
+	if err != nil {
+		return n, err
+	}
+
+	tb.maxTee -= n
```

Comment:
Why do you have to subtract n from maxTee and set it?
This seems to be the state pointer of teeBody, if Write will be called more than once. But on the other hand it's defining the upper limit read from b and write to buffer.